File: ghc-4335c07/nofib/real/fem/inC/assemble_loadvec.c
/* assemble global load vector */

#include "database"

extern float  load_vec[];

assemble_loadvec() {

  int i, dgr, to_point;
  float  px,py,m;

  for (i=1; i<= ndgrs; i=i+1) load_vec[i] = 0.0;

  for (i=1; i<= nplds; i=i+1) {
 
    to_point = load_info[i].to_point;
    px       = load_info[i].px;
    py       = load_info[i].py;
    m        = load_info[i].m;
   
    dgr = node_info[to_point].u;
    if (dgr!=0) load_vec[dgr] = load_vec[dgr] + px;

    dgr = node_info[to_point].v;
    if (dgr!=0) load_vec[dgr] = load_vec[dgr] + py;

    dgr = node_info[to_point].theta;
    if (dgr!=0) load_vec[dgr] = load_vec[dgr] + m;

  }

}

File: ghc-4335c07/nofib/real/fem/Assemble_loadvec.hs
-- Glasow Haskell 0.403 : FINITE ELEMENT PROGRAM V2
-- **********************************************************************
-- *                                                                    *
-- * FILE NAME : assemble_loadvec.hs          DATE : 13-3-1991          *
-- *                                                                    *
-- * CONTENTS : Assemble the global load vector.                        *
-- *                                                                    *
-- **********************************************************************

module Assemble_loadvec (loadvec) where

import Basics
import Vector
import DB_interface
import Degrees

loadvec :: (Array Int Int, Array Int Float) -> Vec Float

loadvec s =
	incrvec initial_value index_value_assoc_s
	where
	initial_value = makevec (ndgrs s) ( \ i -> 0.0 )
	index_value_assoc_s = index_value_assoc s 

index_value_assoc s =
	foldl assemble_s [] [1 .. (nplds s)]
	where
	assemble_s = assemble s 

assemble  s till_now_d ii =
	till_now_d ++ d_this'
	where
	d_this' = filter valid_degree (azip degrees loads)
	valid_degree (dgr , x) = (dgr /=0)
	degrees = getndgr s node
	loads   = [px,py,m]
	(node,px,py,m) = getpld s ii
